File: javascript/src/iterator/value/InsideIteratorValue.ts
//··························································
// Copyright (c) 2023-2026. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {IteratorValue} from "./IteratorValue"

/**
 * An {@link IteratorValue} made to retrieve a consistent value
 * when it is inside a {@link CollectionIterator} range
 *
 * @typeParam T The element type
 * @see GenericCollectionIteratorValue
 * @see IteratorValueOf1
 * @see IteratorValueOf1On2
 * @see IteratorValueOf2On2
 */
export interface InsideIteratorValue<out T, >
    extends IteratorValue<T>,
            IteratorYieldResult<T> {

    /** The index associated to the value */
    readonly index: number


    /** Tell that the {@link InsideIteratorValue iterator value} is able to progress (<i>at the moment of its creation</i>) */
    isDone(): false

    /**
     * Tell that the {@link InsideIteratorValue iterator value} is able to progress (<i>at the moment of its creation</i>)
     *
     * @alias InsideIteratorValue.isDone
     */
    readonly done: false

    /** Tell that the {@link InsideIteratorValue iterator value} is <b>not</b> able to progress (<i>at the moment of its creation</i>) */
    isNotDone(): true


    /** The value to retrieve by the iterator */
    readonly value: T

}

File: java/collection.joookiwi/src/main/java/joookiwi/collection/java/iterator/value/InsideIteratorValue.java
package joookiwi.collection.java.iterator.value;

import joookiwi.collection.java.annotation.Alias;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;

import static joookiwi.collection.java.CommonContracts.ALWAYS_FALSE_0;
import static joookiwi.collection.java.CommonContracts.ALWAYS_TRUE_0;

/// An [IteratorValue] made to retrieve a consistent value
/// when it is inside a [CollectionIterator][joookiwi.collection.java.iterator.CollectionIterator] range
///
/// @param <T> The element type
/// @see GenericCollectionIteratorValue
/// @see IteratorValueOf1
/// @see IteratorValueOf1On2
/// @see IteratorValueOf2On2
@NotNullByDefault
public interface InsideIteratorValue<T extends @Nullable Object>
        extends IteratorValue<T> {

    /// The index associated to the value
    int index();

    /// Tell that the [iterator value][InsideIteratorValue] is able to progress (_at the moment of its creation_)
    @Contract(ALWAYS_TRUE_0) @Override boolean isDone();

    /// Tell that the [iterator value][InsideIteratorValue] is able to progress (_at the moment of its creation_)
    @Alias("isDone") @Contract(ALWAYS_TRUE_0) @Override default boolean done()  { return isDone(); }

    /// Tell that the [iterator value][InsideIteratorValue] is **not** able to progress (_at the moment of its creation_)
    @Contract(ALWAYS_FALSE_0) @Override boolean isNotDone();

}
